src/runtime/weak_ref_count.rs

```diff
@@ -90,30 +60,29 @@ impl WeakRefCount {
             return std::ptr::null_mut();
         }
 
-        let count_or_pointer = self.0.load(Ordering::Relaxed);
+        let mut count_or_pointer = self.0.load(Ordering::Relaxed);
 
         if is_weak_ref(count_or_pointer) {
             return TearOff::from_encoding(count_or_pointer);
         }
 
-        let tear_off = TearOff::new(object, count_or_pointer as _);
-        let encoding: usize =
-            ((tear_off.abi() as usize) >> 1) | (1 << (std::mem::size_of::<usize>() * 8 - 1));
-
         loop {
-            if self
-                .0
-                .compare_exchange_weak(
-                    count_or_pointer,
-                    encoding as _,
-                    Ordering::AcqRel,
-                    Ordering::Relaxed,
-                )
-                .is_ok()
-            {
-                let result: RawPtr = std::mem::transmute(tear_off);
-                TearOff::from_strong_ptr(result).strong_count.add_ref();
-                return result;
+            let tear_off = TearOff::new(object, count_or_pointer as _);
+            let encoding: usize =
+                ((tear_off.abi() as usize) >> 1) | (1 << (std::mem::size_of::<usize>() * 8 - 1));
```

Comment:
Is it necessary to move this inside the loop? This causes the tearoff to be created repeatedly on retry.

Reply:
`tear_off` should only be created once. Thanks for pointing that out!